components/MaskedTransition/src/private/MDCMaskedTransition.m

```diff
@@ -46,83 +46,118 @@ static inline CGFloat LengthOfVector(CGVector vector) {
+    if (!CGRectIsEmpty(finalFrame)) {
+      toView.frame = finalFrame;
+    }
+
+    if (toView.superview == nil) {
+      switch (direction) {
+        case MDMTransitionDirectionForward:
+          [transitionContext.containerView addSubview:toView];
+          break;
+
+        case MDMTransitionDirectionBackward:
+          [transitionContext.containerView insertSubview:toView atIndex:0];
+          break;
+      }
+    }
+  }
+
+  [toView layoutIfNeeded];
+
+  if (direction == MDMTransitionDirectionForward) {
+    return @[from, to];
+  } else {
+    return @[to, from];
+  }
+}
 
 @implementation MDCMaskedTransition {
   UIView *_sourceView;
-  BOOL _shouldSlideWhenCollapsed;
+  MDMTransitionDirection _direction;
 }
 
-- (instancetype)initWithSourceView:(UIView *)sourceView {
+- (instancetype)initWithSourceView:(UIView *)sourceView
+                         direction:(MDMTransitionDirection)direction {
   self = [super init];
   if (self) {
     _sourceView = sourceView;
+    _direction = direction;
   }
   return self;
 }
 
-- (BOOL)canPerformTransitionWithContext:(__unused id<MDMTransitionContext>)context {
-  return _shouldSlideWhenCollapsed ? NO : YES;
-}
-
-#pragma mark - MDMTransitionWithPresentation
+#pragma mark - UIViewControllerAnimatedTransitioning
 
-- (UIModalPresentationStyle)defaultModalPresentationStyle {
-  return UIModalPresentationCustom;
+- (NSTimeInterval)transitionDuration:(id<UIViewControllerContextTransitioning>)transitionContext {
+  return 0.375;
 }
 
-- (UIPresentationController *)presentationControllerForPresentedViewController:(UIViewController *)presented
-                                                      presentingViewController:(UIViewController *)presenting
-                                                          sourceViewController:(__unused UIViewController *)source {
-  MDCMaskedPresentationController *presentationController =
-      [[MDCMaskedPresentationController alloc] initWithPresentedViewController:presented
-                                                      presentingViewController:presenting
-                                                 calculateFrameOfPresentedView:_calculateFrameOfPresentedView];
-  presentationController.sourceView = _sourceView;
-  return presentationController;
-}
-
-- (void)startWithContext:(NSObject<MDMTransitionContext> *)context {
-  MDCMaskedTransitionMotionSpecContext spec = MDCMaskedTransitionMotionSpecForContext(context);
-  if (context.direction == MDMTransitionDirectionForward) {
-    _shouldSlideWhenCollapsed = spec.shouldSlideWhenCollapsed;
-  }
+- (void)animateTransition:(id<UIViewControllerContextTransitioning>)transitionContext {
+  NSArray<UIViewController *> *viewControllers = PrepareTransitionWithContext(transitionContext,
+                                                                              _direction);
+  UIViewController *presentedViewController = viewControllers[1];
```

Comment:
Nit/safety: Do you need a bounds check here?

Reply:
Done.